resolwe_runtime_utils.py

```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
+
+    return "export {}".format(file_path)
+
+
+CHUNK_SIZE = 10000000  # 10 Mbytes
+
+
+class ImportedFormat:
+    """Import destination file format."""
+
+    EXTRACTED = 'extracted'
+    COMPRESSED = 'compressed'
+    BOTH = 'both'
+
+
+def import_file(
+    src,
+    file_name,
+    imported_format=ImportedFormat.BOTH,
+    progress_from=0.0,
+    progress_to=None,
+):
+    """Import file to working directory.
+
+    :param src: Source file path or URL
+    :param file_name: Source file name
+    :param imported_format: Import file format (extracted, compressed or both)
+    :param progress_from: Initial progress value
+    :param progress_to: Final progress value
+    :return: Destination file path (if extracted and compressed, extracted path given)
+    """
+
+    if progress_to is not None:
+        if not isinstance(progress_from, float) or not isinstance(progress_to, float):
+            raise ValueError("progress_from and progress_to must be float")
+
+        if progress_from < 0 or progress_from > 1:
+            raise ValueError("progress_from must be between 0 and 1")
+
+        if progress_to < 0 or progress_to > 1:
+            raise ValueError("progress_to must be between 0 and 1")
+
+        if progress_from >= progress_to:
+            raise ValueError("progress_to must be higher than progress_from")
+
+    print("Importing and compressing {}...".format(file_name))
+
+    def importGz():
+        """Import gzipped file.
+
+        The file_name must have .gz extension.
+        """
+        if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+            with open(file_name[:-3], 'wb') as f_out, gzip.open(src, 'rb') as f_in:
+                shutil.copyfileobj(f_in, f_out, CHUNK_SIZE)
+        else:  # Extracted file not-required
+            # Verify the compressed file.
+            with gzip.open(src, 'rb') as f:
+                try:
+                    while f.read(CHUNK_SIZE) != b'':
+                        pass
+                except Exception:
+                    raise ValueError("invalid gzip file format: {}".format(file_name))
+
+        if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
+            try:
+                shutil.copyfile(src, file_name)
+            except shutil.SameFileError:
+                pass  # Skip copy of downloaded files
+
+        if imported_format == ImportedFormat.COMPRESSED:
+            return file_name
+        else:
+            return file_name[:-3]
+
+    def import7z():
+        """Import compressed file in various formats.
+
+        Supported extensions: .bz2, .zip, .rar, .7z, .tar.gz, and .tar.bz2.
+        """
+        extracted_name, _ = os.path.splitext(file_name)
+        temp_dir = 'temp_{}'.format(extracted_name)
+
+        cmd = '7z x -y -o{} {}'.format(shlex.quote(temp_dir), shlex.quote(src))
+        subprocess.check_call(cmd, shell=True)
```

Comment:
How should I handle the exception?

Reply:
I would print an error message that something went wrong during decompression. We want to fail gracefully in case the file is corrupted. I suggest to do something similar as in the gzip integrity check above.

Reply:
I have decided not to print errors in python processes. We should throw exceptions. The Python process engine should catch the exception and reformat it to process message.

If I print an error (re-error) the python code will continue and likely brake at some lower point. This is then even harder to debug.

Reply:
OK, I'll handle the exception

Reply:
done